### PR TITLE
JOB TITLE API : RDCC -5418

### PR DESCRIPTION
**Pull request by prudhvi-maddineni**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/RDCC-5418

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: src/contractTest/java/uk/gov/hmcts/reform/cwrdapi/StaffReferenceDataProviderTest.java [hunk 1]
```diff
@@ -176,6 +180,16 @@ private CaseWorkerProfile getCaseWorkerProfile(String caseWorkerId) {
...
+    public void retrieveJobTitles() throws JsonProcessingException {
+        List<RoleType> roleTypeList = new ArrayList<>();
+        roleTypeList.add(new RoleType(1L, "Role Description 1"));
+        roleTypeList.add(new RoleType(2L, "Role Description 2"));
+        roleTypeList.add(new RoleType(3L, "Role Description 3"));
+        when(roleTypeRepository.findAll())
+                .thenReturn(roleTypeList);
+    }
+
 }
```
Comment (rajkatla-hmcts) [inline src/contractTest/java/uk/gov/hmcts/reform/cwrdapi/StaffReferenceDataProviderTest.java:195]: have you missed integration tests ? i can't find.
add Integration tests for 
1. add test case for 200 with data and verify the count and their content
2. add test case for 200 empty data and verify the count and empty array

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataJobTitleControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+    List<RoleType> roleType = null;
+
+    @BeforeEach
+    void setUp() {
+        staffRefDataServiceMock = mock(StaffRefDataService.class);
+
+        srJobTitleResponse = StaffRefJobTitleResponse
+                .builder()
+                .jobTitles(Collections.emptyList())
```
Comment (rajkatla-hmcts) [inline src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataJobTitleControllerTest.java:44]: add list of couple of job titles instead of empty list

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/response/StaffRefDataJobTitleResponseTest.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+
+class StaffRefDataJobTitleResponseTest {
+
+    private final StaffRefJobTitleResponse staffRefJobTitleResponse = new StaffRefJobTitleResponse();
+
+    @Test
+    void test_has_mandatory_fields_specified_not_null() {
```
Comment (rajkatla-hmcts) [inline src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/response/StaffRefDataJobTitleResponseTest.java:16]: add test cases for each missing mandatory fields and add assertions for expected exceptions and error messages.

**Review by rajkatla-hmcts — CHANGES_REQUESTED** [review]

**Review by SujeethKumarAri — CHANGES_REQUESTED** [review]

File: src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/client/CaseWorkerApiClient.java [hunk 4]
```diff
@@ -217,4 +220,18 @@ public Response deleteCaseworkerByIdOrEmailPattern(String path, HttpStatus statu
...
+        Response response = getMultipleAuthHeadersInternal(ROLE_STAFF_ADMIN)
+                .get("/refdata/case-worker/job-title")
+                .andReturn();
+        response.then()
+                .assertThat()
+                .statusCode(expectedStatus.value());
+        if (expectedStatus.is2xxSuccessful()) {
+            return response.getBody().as(StaffRefJobTitleResponse.class);
+        } else {
+            return response.getBody().as(ErrorResponse.class);
```
[the remark this answers is not in the document]
Comment (prudhvi-maddineni) [reply src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/client/CaseWorkerApiClient.java]: @rajkatla-hmcts Removed It. I Had added it because to handle the 404 failure response

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataController.java [hunk 5]
```diff
@@ -19,4 +36,60 @@ public class StaffRefDataController {
...
+        if (ObjectUtils.isNotEmpty(roleType)) {
+            List<StaffRefDataJobTitle> refDataJobTitles = roleType.stream()
+                    .map(StaffRefDataJobTitle::new)
+                    .toList();
+            staffRefJobTitleResponseBuilder.jobTitles(refDataJobTitles);
+            log.debug("refDataJobTitles = {}", refDataJobTitles);
+            return   ResponseEntity
+                    .status(200)
+                    .body(staffRefJobTitleResponseBuilder.build());
+        } else {
```
[the remark this answers is not in the document]
Comment (prudhvi-maddineni) [reply src/main/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataController.java]: @rajkatla-hmcts Removed It. I Had added it because to handle the 404 failure response

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataJobTitleControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        responseEntity = ResponseEntity.status(200).body(null);
+        when(staffRefDataServiceMock.getJobTitles())
+                .thenReturn(roleType);
+
+        ResponseEntity<?> actual = staffRefDataController.retrieveJobTitles();
+
+        assertNotNull(actual);
+        verify(staffRefDataServiceMock, times(1))
+                .getJobTitles();
+        assertEquals(responseEntity.getStatusCode(), actual.getStatusCode());
```
[the remark this answers is not in the document]
Comment (prudhvi-maddineni) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataJobTitleControllerTest.java:76]: @rajkatla-hmcts Added it

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataJobTitleControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+
+    @Test
+    void shouldFetchNoJobTitles() {
+        responseEntity = ResponseEntity.status(404).body(null);
```
[the remark this answers is not in the document]
Comment (prudhvi-maddineni) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataJobTitleControllerTest.java]: @rajkatla-hmcts changed the status code 404 to 200

File: src/contractTest/java/uk/gov/hmcts/reform/cwrdapi/StaffReferenceDataProviderTest.java (hunk 1 above)
Comment (prudhvi-maddineni) [reply src/contractTest/java/uk/gov/hmcts/reform/cwrdapi/StaffReferenceDataProviderTest.java:195]: @rajkatla-hmcts As it is new API we are not integrating to any other services. i didn't wrote the Test's. As these 200 cases are covered in functional. Correct me if i am wrong

File: src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefJobTitleFunctionalTest.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    @ToggleEnable(mapKey = mapKey, withFeature = false)
+    void should_fetchJobTitles_403_when_Api_toggled_off() {
+
+        Response response = caseWorkerApiClient.getMultipleAuthHeadersInternal(ROLE_STAFF_ADMIN)
+                .get("/refdata/case-worker/job-title")
+                .andReturn();
+        assertThat(HttpStatus.FORBIDDEN.value()).isEqualTo(response.statusCode());
+        assertThat(response.getBody().asString()).contains(getToggledOffMessage());
+    }
+
```
[the remark this answers is not in the document]
Comment (prudhvi-maddineni) [reply src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefJobTitleFunctionalTest.java:87]: @SujeethKumarAri  Added it

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataController.java [hunk 9]
```diff
@@ -19,4 +36,60 @@ public class StaffRefDataController {
     @Value("${loggingComponentName}")
     private String loggingComponentName;
 
+    @Autowired
+    StaffRefDataService staffRefDataService;
+
+    @ApiOperation(
+            value = "This API is used to retrieve the Job Title's ",
+            notes = "This API will be invoked by Job Title having idam role of staff-admin",
```
[the remark this answers is not in the document]
Comment (prudhvi-maddineni) [reply src/main/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataController.java]: @SujeethKumarAri  Modified

**Review by rajkatla-hmcts — APPROVED** [review]

**Review by SujeethKumarAri — APPROVED** [review]

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataJobTitleControllerTest.java (hunk 2 above)
Comment (prudhvi-maddineni) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataJobTitleControllerTest.java]: @rajkatla-hmcts Modified

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/service/impl/StaffRefDataJobTitleTest.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+    private StaffRefDataServiceImpl staffRefDataService;
+
+    @Test
+    @SuppressWarnings("unchecked")
+    void testFetchJobTitle_All() {
+        when(roleTypeRepository.findAll())
+                .thenReturn(prepareRoleTypeResponse());
+        var roleTypes =  staffRefDataService
+                .getJobTitles();
+        verifyAllRoleTypes(roleTypes);
```
[the remark this answers is not in the document]
Comment (prudhvi-maddineni) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/service/impl/StaffRefDataJobTitleTest.java:35]: @rajkatla-hmcts Added it

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/service/impl/StaffRefDataJobTitleTest.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+    }
+
+    @Test
+    @SuppressWarnings("unchecked")
+    void testFetchJobTitle_one() {
+        when(roleTypeRepository.findAll()).thenReturn(prepareRoleTypeResponse());
+        var staffRefDataUserTypesResponses = (List<RoleType>) staffRefDataService
+                .getJobTitles();
+        assertFalse(verifyCurrentRoleTypes(staffRefDataUserTypesResponses.get(0)));
```
[the remark this answers is not in the document]
Comment (prudhvi-maddineni) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/service/impl/StaffRefDataJobTitleTest.java:50]: @rajkatla-hmcts Added it

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/service/impl/StaffRefDataJobTitleTest.java [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+
+    @Test
+    @SuppressWarnings("unchecked")
+    void testFetchJobTitle_NoDataFound() {
+        var userTypes = prepareRoleTypeResponse();
+        userTypes.clear();
+        when(roleTypeRepository.findAll()).thenReturn(userTypes);
+        var staffRefDataUserTypesResponses = (List<RoleType>) staffRefDataService
+                .getJobTitles();
+        assertTrue(staffRefDataUserTypesResponses.isEmpty());
```
[the remark this answers is not in the document]
Comment (prudhvi-maddineni) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/service/impl/StaffRefDataJobTitleTest.java:63]: @rajkatla-hmcts Added it

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/response/StaffRefDataJobTitleResponseTest.java (hunk 3 above)
Comment (prudhvi-maddineni) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/response/StaffRefDataJobTitleResponseTest.java:16]: @rajkatla-hmcts Added it
